### PR TITLE
Crash recovery

### PR DESCRIPTION
**Pull request by robber-m**
Initial implementation of crash recovery. Plus a lot of merge pain...

Also, added initial implementation for e and E vim-like movement. These movements will need to be enhanced in the future to correctly handle a sequence of word boundary characters like vim does. For example if you hit e with text<cursor_here> ]]) more_text vim would put your cursor on the last non-space word boundary character ). My initial implementation doesn't support this.

@justy989 feel free to change anything I write whenever you want if you have a better or cleaner way of doing things. I won't be offended!

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -135,26 +186,33 @@ int main(int argc, char** argv)
...
+          current_config.view_drawer(buffer_list_head, user_data);
 
           // update the terminal with what we drew
           refresh();
 
           int key = getch();
-
+          if(key == ''){
+               ce_message("reloading config");
+               if(!config_open(&current_config, "ce_config2.so", buffer_list_head, &user_data)){
```
Comment (justy989) [inline main.c]: I don't have a ce_config2.so, can we still use CE_CONFIG here?

File: main.c [hunk 2]
```diff
@@ -35,6 +38,67 @@ void default_destroyer(BufferNode* head, void* user_data);
...
+     return true;
+}
+
+void config_close(Config*config, BufferNode* head, void* user_data){
+     if(!config->so_handle) return;
+     if(config->destroyer) config->destroyer(head, user_data);
+     if(dlclose(config->so_handle)) ce_message("dlclose() failed with error %s", dlerror());
+}
+
+void config_revert(){
```
Comment (justy989) [inline main.c]: This function leaves me dead inside

File: main.c [hunk 3]
```diff
@@ -35,6 +38,67 @@ void default_destroyer(BufferNode* head, void* user_data);
...
+     ce_key_handler* key_handler;
+     ce_view_drawer* view_drawer;
+}Config;
+
+Config stable_config;
+Config current_config;
+
+const Config config_defaults = {NULL, NULL, NULL, default_key_handler, default_view_drawer};
+
+bool config_open(Config* config, const char* path, BufferNode* head, void** user_data){
```
Comment (justy989) [inline main.c]: config_open() doesn't suggest to me that the initialize function is also called, I think I'd lean towards config_open_and_init() or something. Not sure if this is a actual problem

File: main.c (hunk 2 above)
Comment (robber-m) [reply main.c]: hahaha same.

File: main.c (hunk 3 above)
Comment (robber-m) [reply main.c]: I'll do this.

File: main.c (hunk 1 above)
Comment (robber-m) [reply main.c]: sure!

Comment (robber-m) [reply main.c]: done

File: main.c (hunk 2 above)
Comment (robber-m) [reply main.c]: addressed

File: main.c (hunk 3 above)
Comment (robber-m) [reply main.c]: done

**Review by justy989 — APPROVED** [review]
Leave a comment

**Pull request merged**